### ISSUE TITLE
Can I use this project to analyze a whole project or directory?

### ISSUE THREAD

**Issue by chenzhegodspeed** [javaparser/javaparser#2191]
Sometimes I need to figure out call relationships or references for a target method declaration. I know I can search the target method declaration in each compilation unit, but how to get more semantic info?

**Comment by matozoid** [comment]
Hi! Are you asking if the symbol solver produces more information about how it is solving things? The answer would be: no, but it is a good enhancement. But it would be a lot of work.

**Comment by chenzhegodspeed** [comment]
Yes. CompilationUnit is only AST level. I want to know something in semantic level such as call graph or definition and references of a symbol(cross file). I think these information will be more useful.

**Comment by matozoid** [comment]
@ftomassetti ?

**Comment by chenzhegodspeed** [comment]
For example Eclipse-JDT can provide **java mode**l and **AST**.  Eclipse use the **Java Model** to implement some features like Goto Definition, Call hierarchy and so on.  I am not sure is there a similar structure like **Java Model**.

**Comment by edefazio** [comment]
I think I understand what you mean when you say call graphs, (we are talking Afferent and Efferent Dependencies... right?)
[Afferent and Efferent Dependencies](https://stackoverflow.com/questions/15272195/what-is-the-difference-between-afferent-couplings-and-efferent-couplings-of-a-cl)
but I'd probably need to know more information about what you mean by "semantic information", 

The information you need is probably "all there" in the AST, but perhaps "getting to the data" it can be what you are driving at:

I looked at the Eclipse JDT and that's just a DSL built on the AST.  I think what you want is to be able to (for example) get an idea of "who calls who" (A DAG graph) [DAG GRAPH](https://medium.com/@hamzasurti/advanced-data-structures-part-1-directed-acyclic-graph-dag-c1d1145b5e5a)
of your code, it's something you'll have to do manually... it can be done but its a little involved: 
I'd probably start by Walking and collecting all `MethodCallExpr` & `FieldAccessExpr`, etc. that call out)

...but then there's "resolving" the method call signatures with "actual" methods which is much more fun than most people imagine (hello Varargs)

I started down the path of doing some pseudo code, but it got ugly fast... & its not exactly clear what you are asking for.... more clarity would help. (what precisely are you doing? What end goal do you hope to achieve?)...

otherwise a tool like [SONAR](https://dzone.com/articles/working-dependencies-eliminate) might help with this.

**Comment by chenzhegodspeed** [comment]
Yes. I need to get the information of "who calls who".  Every IDE will provide **open declaration**,  **find all references** or **call hierarchy** to user.   I know all information come from AST, but JDT provide some useful interface to obtain it easily(in Eclipse-plugin platform). That's what I need.

**Comment by edefazio** [comment]
OK, got it...
if you want to do this type of functionality in eclipse,  you might want to go ahead and write this in the eclipse JDT
[Building a program with the eclipse JDT](https://www.vogella.com/tutorials/Eclipse/article.html)

So in brief... the AST created by the JavaParser is not the AST created by Eclipse JDT... (they are not common/shared codebases).  They might represent the same thing,(the structure of the code) but they are not "portable" to one another...

actually the AST(or more specifically the "Model" for Java code that Eclipse uses as you type in code for your project) and the model for code that another IDE uses (IntelliJ, Netbeans) are not portable to one another...

FWIW there is an effort to make "Language Servers" that make the API more or less portable... but those are usually limited to help the underlying IDE/editor to do it's job (highlight and code complete) ... (You'd have to take a look at this in more detail)
[Eclipse JDT Language Server](https://github.com/eclipse/eclipse.jdt.ls)

`JavaParser` is great for low level stuff, it can do analysis and other things (refactoring) on a project level, but if you want something closer to the IDE (i.e. a plugin with a UI, etc.) you probably want to build something closer to the tools you are already using.

**Comment by chenzhegodspeed** [comment]
Thanks for your reply. I will modify **JDT** source code to check if it is possible to run outside **Eclipse Platform**. Or I will try **JavaParser** later.

**Comment by matozoid** [comment]
I believe the Spoon project has a parser that uses the Eclipse JDT as the symbol solver.
